[SPARK-46062][SQL] Sync the isStreaming flag between CTE definition and reference

HeartSaVioR opened the pull request:
### What changes were proposed in this pull request?

This PR proposes to sync the flag `isStreaming` from CTE definition to CTE reference. 

The essential issue is that CTE reference node cannot determine the flag `isStreaming` by itself, and never be able to have a proper value and always takes the default as it does not have a parameter in constructor. The other flag `resolved` is handled, and we need to do the same for `isStreaming`.

Once we add the parameter to the constructor, we will also need to make sure the flag is in sync with CTE definition. We have a rule `ResolveWithCTE` doing the sync, hence we add the logic to sync the flag `isStreaming` as well.

### Why are the changes needed?

The bug may impact some rules which behaves differently depending on isStreaming flag. It would no longer be a problem once CTE reference is replaced with CTE definition at some point in "optimization phase", but all rules in analyzer and optimizer being triggered before the rule takes effect may misbehave based on incorrect isStreaming flag.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

HeartSaVioR committed:
[SPARK-46062][SQL] Sync the isStreaming flag between CTE definition and reference

HeartSaVioR commented:
cc. @cloud-fan PTAL, thanks!

HeartSaVioR committed:
empty commit

HeartSaVioR committed:
reduce line

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -862,12 +862,15 @@ case class CTERelationRef(
 
   override lazy val resolved: Boolean = _resolved
 
+  override lazy val isStreaming: Boolean = _isStreaming
```
`isStreaming` is a def, so we can simply put `override val isStreaming`  in the constructor

HeartSaVioR committed:
reflect feedback

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -862,12 +862,15 @@ case class CTERelationRef(
     cteId: Long,
     _resolved: Boolean,
     override val output: Seq[Attribute],
+    _isStreaming: Boolean,
```
```suggestion
    override val isStreaming: Boolean
```
this should work

HeartSaVioR replied:
Ah yes, should be better. Thanks!

HeartSaVioR committed:
Update sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

Co-authored-by: Wenchen Fan <cloud0fan@gmail.com>

HeartSaVioR committed:
fix

HeartSaVioR committed:
update analyzer-results as CTE reference will start to show the flag isStreaming

HeartSaVioR commented:
[29874fc](https://github.com/apache/spark/pull/43966/commits/29874fc07fe0a0272c94c27618062aca289ba72d)

Just fixed other part being broken from the suggestion.

[6840d25](https://github.com/apache/spark/pull/43966/commits/6840d253ed7d6df3f10af3bed91f46eb0ccf8565)

Just updated the golden files - the only diff is that CTE reference node now shows the value of isStreaming flag.

HeartSaVioR commented:
https://github.com/HeartSaVioR/spark/actions/runs/6969118403/job/18967484716

It failed only from test_pandas_map with connection refused error message, which is unrelated.

HeartSaVioR commented:
Thanks, merging to master/3.5/3.4.
